Add support for creating schema in JDBC connector

Praveen2112 opened the pull request:
Fixes #1774

ebyhr commented:
Hmm, I'm not sure if we really want to consolidate properties for dropping schemas and tables. When we add other drop operations (e.g., drop column) to JDBC connector later, this properties may be confusing. 

If we keep current implementation, we need to update existing properties in product test and it would be better to use `@LegacyConfig` for the backward compatibility.

electrum commented:
I’d actually like to get rid of these properties and support file based access control for JDBC connectors.

findepi commented on presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java:
```diff
@@ -306,6 +306,20 @@ private AutoCloseable withTable(String tableName, String tableDefinition)
     }
 
+    @Test
+    public void testCreateSchema()
```
Add the test to `AbstractTestIntegrationSmokeTest` since the functionality is provided by base-jdbc, not postgres connector specific.

kokosing replied:
@Praveen2112 can you please address this comment?

ebyhr replied:
While I agree with adding a test to the class, we still need to override it in each jdbc connector (because DROP SCHEMA implementation was skipped in this PR).

kokosing replied:
That's ok.

findepi commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcClient.java:
```diff
@@ -763,6 +764,35 @@ public TableStatistics getTableStatistics(ConnectorSession session, JdbcTableHan
+    public void dropSchema(JdbcIdentity identity, String schemaName)
+    {
+        if (!getTableNames(identity, Optional.of(schemaName)).isEmpty()) {
+            throw new PrestoException(SCHEMA_NOT_EMPTY, "Schema not empty: " + schemaName);
```
I am not convinced we need this check.
Even if we do the check, a table might be created concurrently (between now and when we issue `DROP SCHEMA`).

What if we do not do this check at all?
I know we won't produce as good err msg, but instead we will produce always the same message (from the underlying database).

cc @electrum

findepi commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcClient.java:
```diff
@@ -763,6 +764,35 @@ public TableStatistics getTableStatistics(ConnectorSession session, JdbcTableHan
+        if (!getTableNames(identity, Optional.of(schemaName)).isEmpty()) {
+            throw new PrestoException(SCHEMA_NOT_EMPTY, "Schema not empty: " + schemaName);
+        }
+        String sql = "DROP SCHEMA " + quoted(schemaName);
```
inline

findepi commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcClient.java:
```diff
@@ -763,6 +764,35 @@ public TableStatistics getTableStatistics(ConnectorSession session, JdbcTableHan
+    @Override
+    public void createSchema(JdbcIdentity identity, String schemaName)
+    {
+        String sql = "CREATE SCHEMA " + quoted(schemaName);
```
inline

kokosing replied:
Some datatabase could have special syntax for that. Possibly we could have a dedicated method to to generate drop and create schema SQLs that implementation of JDBC based connector could override.

findepi replied:
`createSchema` is so simple that `BaseJdbcClient` subclass can override the whole method, if needed.

findepi commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcClient.java:
```diff
@@ -763,6 +764,35 @@ public TableStatistics getTableStatistics(ConnectorSession session, JdbcTableHan
+    }
+
+    @Override
+    public void dropSchema(JdbcIdentity identity, String schemaName)
```
We need a feature toggle for this.
Even if long term plan is to support file-based authZ, we need backward compatibility.

I am not sure it's safe to enable this by default.
A schema might be associated with some configuration and drop/recreate may effectively change this configuration. (Like "location" in Hive, or maybe tablespace in a traditional database).
I'd rather have it off by default.

cc @ebyhr @electrum

kokosing commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcClient.java:
```diff
@@ -763,6 +764,35 @@ public TableStatistics getTableStatistics(ConnectorSession session, JdbcTableHan
+    @Override
+    public void dropSchema(JdbcIdentity identity, String schemaName)
+    {
+        if (!getTableNames(identity, Optional.of(schemaName)).isEmpty()) {
```
Schema may contain other objects (not only tables and views). Also schema could be used in other objects (like being used in view definition) and so underlying database could prevent removal of the schema here. In general it looks we are not able to verify if schema is safe to be removed.  Hence I would remove this check.

Praveen2112 commented:
@electrum , @ebyhr , @kokosing , @findepi I have modified this PR to add support only for creation of schema for jdbc based connectors while dropping of schema can be started once we have added support for file based authZ

kokosing commented on presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java:
```diff
@@ -207,6 +207,14 @@ public void testInsertIntoNotNullColumn()
+    public void testCreateSchema()
+            throws SQLException
+    {
+        assertUpdate("CREATE SCHEMA schema1");
```
- run `SHOW SCHEMAS`
- s/schema1/test_create_schema

Praveen2112 commented:
@kokosing Thanks for your insights. Have applied your comments

ebyhr commented:
@kokosing Shall we merge this PR?

kokosing commented:
> Shall we merge this PR?

@ebyhr  There is one pending comment: https://github.com/prestosql/presto/pull/1874/files#r339430588

Praveen2112 commented:
@kokosing Have applied that comment.

kokosing commented on presto-base-jdbc/src/test/java/io/prestosql/plugin/jdbc/TestJdbcIntegrationSmokeTest.java:
```diff
@@ -113,17 +113,29 @@ public void testUnknownTypeAsVarchar()
-                    "VALUES (1, NULL), (2, 'POINT (7 52)'), (3, NULL)");
+                convertToVarcharUnsupportedTypes,
+                "SELECT * FROM " + table.getName(),
+                "VALUES (1, NULL), (2, 'POINT (7 52)'), (3, NULL)");
```
undo

kokosing commented on presto-base-jdbc/src/test/java/io/prestosql/plugin/jdbc/TestJdbcIntegrationSmokeTest.java:
```diff
@@ -113,17 +113,29 @@ public void testUnknownTypeAsVarchar()
-                .setCatalogSessionProperty("jdbc", UNSUPPORTED_TYPE_HANDLING, unsupportedTypeHandling.name())
-                .build();
+            .setCatalogSessionProperty("jdbc", UNSUPPORTED_TYPE_HANDLING, unsupportedTypeHandling.name())
+            .build();
```
undo formatting

kokosing commented on presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java:
```diff
@@ -213,15 +213,28 @@ public void testInsertIntoNotNullColumn()
+
     @Test
     public void testColumnComment()
-            throws Exception
```
undo

kokosing commented on presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java:
```diff
@@ -213,15 +213,28 @@ public void testInsertIntoNotNullColumn()
+    }
+
+    @Override
+    protected void cleanUpSchema()
```
cleanup should take schema parameter name

kokosing commented on presto-testing/src/main/java/io/prestosql/testing/AbstractTestIntegrationSmokeTest.java:
```diff
@@ -231,16 +247,29 @@ private MaterializedResult getExpectedOrdersTableDescription(boolean dateSupport
+        assertQuery("SHOW SCHEMAS LIKE 'test_create_schema'", "VALUES 'test_create_schema'");
+        cleanUpSchema();
+        if (canDropSchema()) {
+            assertUpdate("DROP SCHEMA test_create_schema");
```
add dedicated test for drop which is not run when connector does not support that

kokosing replied:
schema is already dropped here

kokosing commented on presto-testing/src/main/java/io/prestosql/testing/AbstractTestIntegrationSmokeTest.java:
```diff
@@ -243,4 +261,26 @@ private MaterializedResult getExpectedOrdersTableDescription(boolean dateSupport
+        skipTestUnless(canCreateSchema());
+        assertUpdate("CREATE SCHEMA test_create_schema");
+        assertQuery("SHOW SCHEMAS LIKE 'test_create_schema'", "VALUES 'test_create_schema'");
+        cleanUpSchema(ImmutableList.of("test_create_schema"));
```
cleanUpSchema  -> cleanupSchemas

kokosing commented on presto-testing/src/main/java/io/prestosql/testing/AbstractTestIntegrationSmokeTest.java:
```diff
@@ -41,6 +45,20 @@ protected boolean isParameterizedVarcharSupported()
+
+    protected void cleanUpSchemas(List<String> schemaNames)
+            throws SQLException
+    {}
```
throw an exception when this method is not implemented. So in case someone return `false` from `canDropSchema` we don't leave a trash behind.

kokosing commented on presto-testing/src/main/java/io/prestosql/testing/AbstractTestIntegrationSmokeTest.java:
```diff
@@ -243,4 +261,28 @@ private MaterializedResult getExpectedOrdersTableDescription(boolean dateSupport
+            throws SQLException
+    {
+        skipTestUnless(canCreateSchema());
+        assertUpdate("CREATE SCHEMA test_create_schema");
```
test could be run concurrently and single connector can have multiple implementation of this abstract class. Using hardcoded name will make this test to be flaky. Please use `io.prestosql.testing.sql.TestTable#randomTableSuffix` to make this schema name unique. Some for test below.

Currently you are using same schema name in this test and in `testDropSchema` so they are flaky even if there is a single implementation of this class and concurrent tests execution.

kokosing commented on presto-testing/src/main/java/io/prestosql/testing/AbstractTestIntegrationSmokeTest.java:
```diff
@@ -243,4 +261,28 @@ private MaterializedResult getExpectedOrdersTableDescription(boolean dateSupport
+    @Test
+    public void testDropSchema()
+    {
+        skipTestUnless(canDropSchema());
```
make sure that schema does not exist before creation, that it doesn't exists after deletion

kokosing replied:
Also verify that you are unable to create schema when it exists, and to drop when it does not exist (in both tests).

kokosing commented on presto-testing/src/main/java/io/prestosql/testing/AbstractTestIntegrationSmokeTest.java:
```diff
@@ -243,4 +261,28 @@ private MaterializedResult getExpectedOrdersTableDescription(boolean dateSupport
+    public void testCreateSchema()
+            throws SQLException
+    {
+        skipTestUnless(canCreateSchema());
```
make sure that schema does not exist before creation, that it doesn't exists after deletion
